### PR TITLE
ability to make bundle with required params

### PR DESCRIPTION
**Pull request by leelavg**
ref: #110

Signed-off-by: Leela Venkaiah G <lgangava@redhat.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -30,6 +30,8 @@ endif
 OS = $(shell go env GOOS)
 ARCH = $(shell go env GOARCH)
 
+OPERATOR_NAMESPACE ?=
```
Comment (nb-ohad) [inline Makefile:33]: Why do we need `OPERATOR_NAMESPACE` when we already have `NAMESPACE`?

File: Makefile [hunk 2]
```diff
@@ -30,6 +30,8 @@ endif
...
+REPLACES ?=
```
Comment (nb-ohad) [inline Makefile:34]: Please move this definition immediately after `VERSION` definition (line2) as they are tightly coupled. Also, this should have a default value of the current version we are replacing

File: Makefile [hunk 3]
```diff
@@ -179,11 +181,17 @@ endif
...
 	operator-sdk generate kustomize manifests -q
 	cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
+	cd config/manifests/bases && sed -e "s/@REPLACES@/$(REPLACES)/" -e "s/@VERSION@/$(VERSION)/" kustomization-base.yaml > kustomization.yaml
+	cd config/manifests && $(KUSTOMIZE) edit set namespace $(OPERATOR_NAMESPACE)
 	$(KUSTOMIZE) build config/manifests | operator-sdk generate bundle -q --extra-service-accounts prometheus-k8s --overwrite --version $(VERSION) $(BUNDLE_METADATA_OPTS) $(BUNDLE_FLAGS)
 	cp config/metadata/* $(OUTPUT_DIR)/metadata/
 	operator-sdk bundle validate $(OUTPUT_DIR)
+else
+	@echo Variables 'REPLACES' and 'OPERATOR_NAMESPACE' are mandatory for generating bundle
+endif
```
Comment (nb-ohad) [inline Makefile:194]: You can skip this check as we the above requests will guarantee default values for both

File: config/manifests/bases/kustomization-base.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+kind: Kustomization
+resources:
+- ocs-osd-deployer.clusterserviceversion.yaml
+commonAnnotations:
+  olm.skipRange: '>=0.0.1 <@VERSION@'
+patches:
+- patch: '[{"op": "replace", "path": "/spec/replaces", "value": "@REPLACES@"}]'
+  target:
+    kind: ClusterServiceVersion
+    name: ocs-osd-deployer.v0.0.0
```
Comment (nb-ohad) [inline config/manifests/bases/kustomization-base.yaml:11]: Why do we need this `-base` file if we can already work on the actual file?

File: Makefile [hunk 5]
```diff
@@ -179,11 +181,17 @@ endif
 # Generate bundle manifests and metadata, then validate generated files.
 .PHONY: bundle
 bundle: manifests kustomize
+ifneq ($(and $(REPLACES),$(OPERATOR_NAMESPACE)),)
 	operator-sdk generate kustomize manifests -q
 	cd config/manager && $(KUSTOMIZE) edit set image controller=$(IMG)
+	cd config/manifests/bases && sed -e "s/@REPLACES@/$(REPLACES)/" -e "s/@VERSION@/$(VERSION)/" kustomization-base.yaml > kustomization.yaml
```
Comment (nb-ohad) [inline Makefile:187]: @leelavg I think this change came from a misunderstanding of the intant of my prev request. 
I would prefer to edit the actual file and not create a template to be edited by `sed`. Please use `yq` or any other tool (even `kustomize` itself if you find it easier) to edit the actual patch file.

File: Makefile [hunk 6]
```diff
@@ -179,11 +181,17 @@ endif
...
+	cd config/manifests && $(KUSTOMIZE) edit set namespace $(OPERATOR_NAMESPACE)
```
Comment (nb-ohad) [inline Makefile:188]: Why is this needed? NAMESPACE is already taken care of in `config/default/kustomization.yaml` and I would like to have a single source of truth regarding the namespace.

I think it will be useful to use this command to set the namespace that is already defined there.

File: config/manifests/kustomization.yaml [hunk 7]
```diff
@@ -1,4 +1,8 @@
 resources:
+- bases
 - ../default
 - ../samples
 - ../scorecard
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+namespace: default
```
Comment (nb-ohad) [inline config/manifests/kustomization.yaml:8]: Why do we need this change? It works perfectly without it

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: Makefile (hunk 1 above)
Comment (leelavg) [reply Makefile]: removed.

File: Makefile (hunk 2 above)
Comment (leelavg) [reply Makefile]: done.

File: config/manifests/bases/kustomization-base.yaml (hunk 4 above)
Comment (leelavg) [reply config/manifests/bases/kustomization-base.yaml]: removed

File: Makefile (hunk 6 above)
Comment (leelavg) [reply Makefile]: removed, not required.

File: config/manifests/kustomization.yaml (hunk 7 above)
Comment (leelavg) [reply config/manifests/kustomization.yaml]: removed.

File: Makefile (hunk 5 above)
Comment (leelavg) [reply Makefile]: - Let me explain why I did what I did and I don't want to use neither `sed` nor  introduce a new one `yq`

> 	cd config/manifests/bases && \
- normal cd op

>		rm -rf kustomization.yaml && \
- we can only update below resources/fields with `kustomize edit set`
```
image
label
nameprefix
namespace
namesuffix
replicas
```
- other op `kustomize edit add` will add to the existing resource patch file and since we can't override existing field in the patch file, we are removing it

>		$(KUSTOMIZE) create --resources $(BUNDLE_PACKAGE).clusterserviceversion.yaml && \
- this creates a patch file with resource section `$(BUNDLE_PACKAGE).clusterserviceversion.yaml` on which patches has to be applied not the CSV itself

>		$(KUSTOMIZE) edit add annotation --force 'olm.skipRange':">=0.0.1 <$(VERSION)" && \
- first change we want in a csv, i.e, to update the skipRange

>		$(KUSTOMIZE) edit add patch --name $(BUNDLE_PACKAGE).v0.0.0 --kind ClusterServiceVersion \
		--patch '[{"op": "replace", "path": "/spec/replaces", "value": "$(REPLACES)"}]'
- unlike `kustomize edit add annotation` we can't directly target the fields under spec so we are adding a strategic merge patchable json object

---
- Against sed:
1. If we want inplace replace, the resulting expression will be cryptic and may not be cross-platform
- Against yq: 
1. extra dependency, the field which we want to edit is a json obj inside string and it'll more or less result in same format as kustomize minus 2 steps (remove and create of kustomization.yaml)
2. those two steps may not cause us much friction while generating bundle and can get away with one less installable dependency

---
- All I'm doing is to clarify as much as possible and correct any of my misunderstandings how the tools work and I believe these changes doesn't need a lot of discussions

**Comment by leelavg** [comment]
- rebased

**Comment by leelavg** [comment]
- rebased, @nb-ohad could you pls take a re-look at this PR, thanks.

**Review by bindrad — APPROVED** [review]

File: Makefile [hunk 8]
```diff
@@ -33,6 +34,8 @@ endif
 OS = $(shell go env GOOS)
 ARCH = $(shell go env GOARCH)
 
+BUNDLE_PACKAGE ?= ocs-osd-deployer
```
Comment (nb-ohad) [inline Makefile:37]: @leelavg Why are we allowing the invoker to override the bundle package? Also if someone will run it with a different value the result will be unpredictable, as it will not create a new CSV with the new name, it will just create a patch for a resource that does not exist.

I would suggest just embedding the name `ocs-osd-deployer` in the two commands that use it.

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: Makefile (hunk 8 above)
Comment (leelavg) [reply Makefile]: done.

**Review by nb-ohad — APPROVED** [review]
